File: Chapter 3 (Searching problems)/Informed Search/Python/astar.py
import heapq

# A* with weighted adjacency matrix graph and heuristic vector h
def astar(graph, start, goal, h):
    n=len(graph)
    openq=[]
    heapq.heappush(openq,(h[start],0,start))
    parent=[-1]*n
    g=[float('inf')]*n
    g[start]=0
    closed=[False]*n
    while openq:
        f,gv,u=heapq.heappop(openq)
        if closed[u]:
            continue
        if u==goal:
            path=[]
            while u!=-1:
                path.append(u)
                u=parent[u]
            return list(reversed(path))
        closed[u]=True
        for v in range(n):
            w=graph[u][v]
            if w!=-1:
                ng=g[u]+w
                if ng<g[v]:
                    g[v]=ng
                    parent[v]=u
                    heapq.heappush(openq,(ng+h[v],ng,v))
    return []

if __name__=="__main__":
    names=['A','B','C','D']
    n=len(names)
    g=[[-1]*n for _ in range(n)]
    def add(u,v,w): g[u][v]=w
    add(0,1,1); add(0,2,4); add(1,2,2); add(1,3,5); add(2,3,1)
    h=[7,6,2,0]
    path=astar(g,0,3,h)
    print([names[i] for i in path])

File: Chapter 3 (Searching problems)/Informed Search/C++/astar.cpp
#include <bits/stdc++.h>
#include <bits/stdc++.h>
using namespace std;

// A* Search on adjacency matrix with non-negative weights and heuristic vector.
vector<int> astar(const vector<vector<int>> &g, const vector<int> &h, int s, int t)
{
    int n = (int)g.size();
    using T = tuple<int, int, int>; // f, g, node
    priority_queue<T, vector<T>, greater<T>> openq;
    openq.push({h[s], 0, s});
    vector<int> gval(n, INT_MAX), par(n, -1);
    gval[s] = 0;
    vector<char> closed(n, 0);
    while (!openq.empty())
    {
        auto [f, gv, u] = openq.top();
        openq.pop();
        if (closed[u])
            continue;
        if (u == t)
        {
            vector<int> path;
            for (int x = u; x != -1; x = par[x])
                path.push_back(x);
            reverse(path.begin(), path.end());
            return path;
        }
        closed[u] = 1;
        for (int v = 0; v < n; ++v)
        {
            int w = g[u][v];
            if (w != -1)
            {
                int ng = gval[u] + w;
                if (ng < gval[v])
                {
                    gval[v] = ng;
                    par[v] = u;
                    openq.push({ng + h[v], ng, v});
                }
            }
        }
    }
    return {};
}

int main()
{
    vector<string> names = {"A", "B", "C", "D"};
    int n = (int)names.size();
    vector<vector<int>> g(n, vector<int>(n, -1));
    auto add_edge = [&](int u, int v, int w)
    { g[u][v] = w; };
    add_edge(0, 1, 1);
    add_edge(0, 2, 4);
    add_edge(1, 2, 2);
    add_edge(1, 3, 5);
    add_edge(2, 3, 1);
    vector<int> h = {7, 6, 2, 0};

    auto p = astar(g, h, 0, 3);
    for (size_t i = 0; i < p.size(); ++i)
    {
        if (i)
            cout << " ";
        cout << names[p[i]];
    }
    cout << "\n";
}
